marimo/_save/save.py

```diff
@@ -30,42 +39,313 @@
+
+    return _cache_base(
+        _fn,
+        maxsize=maxsize,
+        pin_modules=pin_modules,
+        frame_offset=1,
+    )
+
+
 class persistent_cache(object):
-    """Context block for cache lookup of a block of code.
+    """Save variables to disk and restore them thereafter.
 
-    Example usage:
+    The `mo.persistent_cache` context manager lets you delimit a block of code
+    in which variables will be cached to disk when they are first computed. On
+    subsequent runs of the cell, if marimo determines that this block of code
+    hasn't changed and neither has its ancestors, it will restore the variables
+    from disk instead of re-computing them, skipping execution of the block
+    entirely.
 
-    >>> with persistent_cache(name="my_cache"):
-    >>>     variable = expensive_function() # This will be cached.
+    Restoration happens even across notebook runs, meaning you can use
+    `mo.persistent_cache` to make notebooks start *instantly*, with variables
+    that would otherwise be expensive to compute already materialized in
+    memory.
 
-    For an implementation sibling regarding the block skipping, see `withhacks`
-    in pypi.
+    **Usage.**
 
-    NB: Since context abuses sys frame trace, this may conflict with debugging
-    tools that also use sys.settrace.
+    ```python
+    with persistent_cache(name="my_cache"):
+        variable = expensive_function()  # This will be cached to disk.
+        print("hello, cache")  # this will be skipped on cache hits
+    ```
+
+    In this example, `variable` will be cached the first time the block
+    is executed, and restored on subsequent runs of the block. If cache
+    conditions are hit, the contents of `with` block will be skipped on
+    execution. This means that side-effects such as writing to stdout and
+    stderr will be skipped on cache hits.
+
+    For function-level memoization, use `@mo.cache` or `@mo.lru_cache`.
+
+    Note that `mo.state` and `UIElement` changes will also trigger cache
+    invalidation, and be accordingly updated.
+
+    **Warning.** Since context abuses sys frame trace, this may conflict with
+    debugging tools or libraries that also use `sys.settrace`.
+
+    **Args**:
+
+    - `name`: the name of the cache, used to set saving path- to manually
+      invalidate the cache, change the name.
+    - `save_path`: the folder in which to save the cache, defaults to
+      ".marimo/cache"
+    - `pin_modules`: if True, the cache will be invalidated if module versions
+      differ between runs, defaults to False.
     """
 
     def __init__(
         self,
-        *,
-        save_path: str = "outputs",
         name: str,
+        *,
+        save_path: str = ".marimo/cache",
         pin_modules: bool = False,
         _loader: Optional[Loader] = None,
     ) -> None:
+        # For an implementation sibling regarding the block skipping, see
+        # `withhacks` in pypi.
         self.name = name
+        save_path = save_path
```

Comment:
@dmadisetti Did you want to update the `save_path` to append the notebook name? Mainly to make it easier for users to manage their caches.

If so, then the signature could be updated to:

`save_path: str | None = None`

and then `if save_path is None`, `save_path = ".marimo/cache/{notebook-name}/name"`

?

Reply:
Hmm. I was thinking about this, but:

1. There should never be a cache collision + the user can add the specificity if they want
2. The directory is still local to the notebook 
3. It allows speedup between notebooks

But 3. Might be hard to intentionally trigger, and the notebook name makes things easy to manually clean up. I think adding notebook name seems reasonable but:

 - What happens on notebook rename
 - is the only mechanism for cleanup manually deleting files?

Reply:
> What happens on notebook rename

I think the user would have to manually rename the cache :/ 

> is the only mechanism for cleanup manually deleting files?

Did you have another idea in mind?

Reply:
> I think the user would have to manually rename the cache :/

Not specifying notebook is a quick way around that

> Did you have another idea in mind?

I think a remote LRU cache would be cool, but nothing locally

Reply:
> Not specifying notebook is a quick way around that

Yea -- i think it ends up being what do we value more:

1. Not having to manually rename the cache on notebook rename?
2. Making it easy to manually clear the cache for a specific notebook?

I was leaning toward 2 as being more valuable, since I'm assuming the user might be caching big objects. But I'm unsure. Thoughts?

Reply:
I think cell level caching should be under `<dir>`/`cache`/`<notebook name>` for sure.

`name` is required for persistent_cache, so `<dir>`/`cache`/`my_large_cache_object` or `<dir>`/`persistent_cache`/`my_large_cache_object` (to separate the 2 caches) seems reasonable to me. The user could always specify `name=notebook_name/large_cache`, which will create the sub directory.

Reply:
Fair enough! Keep as is then.